cmd/drone-buildx-gcr/main.go

```diff
@@ -1,63 +1,159 @@
 package main
 
 import (
+	"context"
 	"encoding/base64"
+	"log"
 	"os"
 	"path"
+	"strconv"
 	"strings"
 
 	docker "github.com/drone-plugins/drone-buildx"
+	"github.com/drone-plugins/drone-buildx-gcr/internal/gcp"
 	"github.com/joho/godotenv"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
 )
 
-// gcr default username
-const username = "_json_key"
+type Config struct {
+	Repo             string
+	Registry         string
+	Password         string
+	WorkloadIdentity bool
+	Username         string
+	AccessToken      string
+}
 
-func main() {
-	// Load env-file if it exists first
+type staticTokenSource struct {
+	token *oauth2.Token
+}
+
+func (s *staticTokenSource) Token() (*oauth2.Token, error) {
+	return s.token, nil
+}
+
+func loadConfig() Config {
+	// Default username
+	username := "_json_key"
+	var config Config
+
+	// Load env-file if it exists
 	if env := os.Getenv("PLUGIN_ENV_FILE"); env != "" {
-		godotenv.Load(env)
+		if err := godotenv.Load(env); err != nil {
+			log.Fatalf("Error loading .env file: %v", err)
+		}
 	}
 
-	var (
-		repo     = getenv("PLUGIN_REPO")
-		registry = getenv("PLUGIN_REGISTRY")
-		password = getenv(
+	idToken := getenv("PLUGIN_OIDC_TOKEN_ID")
+	projectId := getenv("PLUGIN_PROJECT_NUMBER")
+	poolId := getenv("PLUGIN_POOL_ID")
+	providerId := getenv("PLUGIN_PROVIDER_ID")
+	serviceAccountEmail := getenv("PLUGIN_SERVICE_ACCOUNT_EMAIL")
+
+	if idToken != "" && projectId != "" && poolId != "" && providerId != "" && serviceAccountEmail != "" {
+		federalToken, err := gcp.GetFederalToken(idToken, projectId, poolId, providerId)
+		if err != nil {
+			logrus.Fatalf("PLUGIN_OIDC_TOKEN_ID: %s, PLUGIN_PROJECT_NUMBER: %s, PLUGIN_POOL_ID: %s, PLUGIN_PROVIDER_ID: %s, PLUGIN_SERVICE_ACCOUNT_EMAIL: %s", idToken, projectId, poolId, providerId, serviceAccountEmail)
+			logrus.Fatalf("Error (getFederalToken): %s", err)
+		}
+		accessToken, err := gcp.GetGoogleCloudAccessToken(federalToken, serviceAccountEmail)
+		if err != nil {
+			logrus.Fatalf("Error (getGoogleCloudAccessToken): %s", err)
+		}
+		config.AccessToken = accessToken
+	} else {
```

Comment:
Make this a simple if else.

`If workload identity flow:`
`    set username, accesstoken, repo, registry`
`else:`
`    set username, password, repo, registry`

Reply:
Looks better now. Thanks for the refactor